Feat/ota 497/more ecu events

lbonn opened the pull request:
Will still move a bit. We'll have to keep an id circulating between these events to link them with each others.

patrickvacek commented on tests/httpfake.h:
```diff
@@ -134,6 +134,9 @@ class HttpFake : public HttpInterface {
         return HttpResponse(Utils::readFile("tests/test_data/cred.p12"), 200, CURLE_OK, "");
+      } else if (url == "tst149/events") {
+        ++events_seen;
+        return HttpResponse("", 200, CURLE_OK, "");
```
The test149 section is only used by the `ProvisionOnServer` test in uptane_test. It might be worth adding checks in that test that `events_seen` has the expected values as we check for the other event counts.

lbonn commented on src/libaktualizr/primary/sotauptaneclient.cc:
```diff
@@ -708,12 +708,30 @@ bool SotaUptaneClient::downloadImages(const std::vector<Uptane::Target> &targets
+  bool success = false;
+  if (uptane_fetcher->fetchVerifyTarget(target)) {
+    success = true;
+  }
```
Thanks!

lbonn commented on src/libaktualizr/primary/sotauptaneclient.cc:
```diff
@@ -814,18 +832,24 @@ void SotaUptaneClient::uptaneInstall(const std::vector<Uptane::Target> &updates)
       storage->storeInstallationResult(result);
-      sendEvent<event::InstallStarted>(uptane_manifest.getPrimaryEcuSerial());
+      report_queue->enqueue(
+          std_::make_unique<EcuInstallationCompletedReport>(uptane_manifest.getPrimaryEcuSerial(), false));
```
Yes you're right. Maybe the right solution would be to ditch the boolean and start using multiple codes (failed, ok, already_installed...)

lbonn commented on src/libaktualizr/primary/sotauptaneclient.cc:
```diff
@@ -1007,11 +1031,19 @@ void SotaUptaneClient::sendMetadataToEcus(const std::vector<Uptane::Target> &tar
+    report_queue->enqueue(std_::make_unique<EcuInstallationStartedReport>(secondary.getSerial()));
+    bool ret = secondary.sendFirmware(data);
+    report_queue->enqueue(std_::make_unique<EcuInstallationCompletedReport>(secondary.getSerial(), ret));
+    sendEvent<event::InstallTargetComplete>(secondary.getSerial(), ret);
```
Yes there is a big overlap. For now I think it would be simpler to keep them decoupled, we can see how it goes :)

patrickvacek replied:
That's fine. But it does seem like in the long term the goals are the same: to be able to show download and installation start/progress/success/failure either in an HMI via the API or on the server.

lbonn commented:
I've addressed some of your comments @patrickvacek but left some over that would need a bit more work.

Also added a last change to flush the reportqueue when the program exits. Otherwise, we could sometimes lose events, especially when using individual steps through the command line.

patrickvacek commented:
> but left some over that would need a bit more work.

That's fine, but do you mind elaborating which things will still need to be addressed at a later time? 

> Also added a last change to flush the reportqueue when the program exits. Otherwise, we could sometimes lose events, especially when using individual steps through the command line.

Good call, I was always aware of that danger and it needed to be done. Thanks!
